### PR TITLE
episodes, experiences design

### PR DESCRIPTION
**Pull request by StanislavVolodarskiy**

### CODE REVIEW

File: algorithms/da3c/da3c_model.py [hunk 1]
```diff
@@ -16,15 +16,15 @@ def build_graph(self):
...
 
-        fc = layer.Dense(graph.Flatten(conv), 256, activation=graph.Relu)
+        fc = layer.Dense(layer.Flatten(conv), 256, activation=graph.Relu)
 
         actor = layer.Dense(fc, da3c_config.config.action_size, activation=graph.Softmax)
         critic = layer.Dense(fc, 1)
 
         self.state = state
         self.actor = graph.Softmax(actor)
         self.critic = graph.Flatten(critic)
```
Comment (4SkyNet) [inline algorithms/da3c/da3c_model.py]: Good

Comment (4SkyNet) [inline algorithms/da3c/da3c_model.py]: Good

File: algorithms/da3c/da3c_model.py [hunk 2]
```diff
@@ -70,7 +74,7 @@ def build_graph(self):
 
         # Expose public API
         self.op_assign_weights = self.Op(sg_assign_weights, weights=ph_weights)
```
Comment (4SkyNet) [inline algorithms/da3c/da3c_model.py]: seems like a lot of `assign` issues and related code for this, which also not so obvious to user/developer
We've to expose assign interface, but it looks like more implicit and we probably can hide it a little a bit

File: relaax/algorithms/da3c/da3c_model.py [hunk 3]
```diff
@@ -31,27 +31,24 @@ def build_graph(self):
...
-        ph_gradients = graph.Placeholders(variables=sg_weights)
         sg_learning_rate = da3c_graph.LearningRate(sg_global_step)
         sg_optimizer = graph.RMSPropOptimizer(
             learning_rate=sg_learning_rate,
             decay=da3c_config.config.RMSProp.decay,
             momentum=0.0,
             epsilon=da3c_config.config.RMSProp.epsilon
         )
-        sg_apply_gradients = graph.ApplyGradients(sg_optimizer, sg_weights, ph_gradients)
+        sg_gradients = layer.Gradients(sg_weights, optimizer=sg_optimizer)
```
Comment (4SkyNet) [inline relaax/algorithms/da3c/da3c_model.py]: To my mind `Gradients` is more `Network` stuff, than ordinary `layer`, since that:
it's better to move them as `Network` or `Weights` part if it possible

File: relaax/algorithms/da3c/da3c_model.py [hunk 4]
```diff
@@ -60,25 +57,18 @@ class AgentModel(subgraph.Subgraph):
...
         sg_network = Network()
-        ph_state = sg_network.state
 
-        ph_action = graph.Placeholder(np.int32, shape=(None, ))
-        ph_value = graph.Placeholder(np.float32, shape=(None, ))
-        ph_discounted_reward = graph.Placeholder(np.float32, shape=(None, ))
-        sg_loss = da3c_graph.Loss(ph_state, ph_action, ph_value,
-                ph_discounted_reward, sg_network.actor, sg_network.critic)
-        sg_gradients = graph.Gradients(sg_loss, sg_network.weights)
+        sg_loss = da3c_graph.Loss(sg_network.actor, sg_network.critic)
```
Comment (4SkyNet) [inline relaax/algorithms/da3c/da3c_model.py]: All `Losses` is dependent wrt appropriate graph as I understand for now.
It's better to them into auxiliary `Losses` pack with choice between `A3C_Loss`, `PG_Loss` and so on
with the relevant pop-up of parameter's help

File: relaax/algorithms/da3c/da3c_model.py [hunk 5]
```diff
@@ -60,25 +57,18 @@ class AgentModel(subgraph.Subgraph):
...
-        self.op_get_action_and_value = self.Ops(sg_network.actor, sg_network.critic, state=ph_state)
-        self.op_compute_gradients = self.Op(sg_gradients,
-            state=ph_state,
-            action=ph_action,
-            value=ph_value,
-            discounted_reward=ph_discounted_reward
-        )
+        self.op_get_action_and_value = self.Ops(
+                sg_network.actor, sg_network.critic, state=sg_network.state)
+        self.op_compute_gradients = self.Op(sg_gradients.calculate,
```
Comment (4SkyNet) [inline relaax/algorithms/da3c/da3c_model.py]: `Gradients` computation is `Loss` dependent, so 
-> can we suppress these `3` params wrt to current `loss` graph by simple:
```python
self.op_compute_gradients = self.Op(sg_gradients.calculate, state=sg_network.state, loss=sg_loss)
```

**Review by 4SkyNet — COMMENTED** [review]
There are small reflections and remarks
